File: python/src/aoc_2024_04.py
"""Advent of Code 2024 - Day 4."""

from sys import stdin

from coord import Coord


class LetterGrid:
    """Hold our letter grid and methods."""

    def __init__(self, lines: list[str]) -> None:
        self.height = len(lines)
        self.width = len(lines[0])
        self.lines = lines
        for line in lines:
            assert len(line) == self.width

    def letter(self, c: Coord) -> str:
        """Return letter at given coordinate.

        Return space if out of bounds.
        """
        if c.x < 0 or c.x >= self.width or c.y < 0 or c.y >= self.height:
            return " "
        return self.lines[c.y][c.x]

    def check_xmas(self, x_coord: Coord, direction: Coord) -> bool:
        """Check to see for XMAS in given direction."""
        c = x_coord + direction
        for letter in "MAS":
            if self.letter(c) != letter:
                return False
            c += direction
        return True

    def count_xmas(self) -> int:
        """Count occurrences of XMAS."""
        count = 0
        for x in range(self.width):
            for y in range(self.height):
                c = Coord(x, y)
                if self.letter(c) == "X":
                    for d in (
                        Coord(0, 1),
                        Coord(1, 1),
                        Coord(1, 0),
                        Coord(1, -1),
                        Coord(0, -1),
                        Coord(-1, -1),
                        Coord(-1, 0),
                        Coord(-1, 1),
                    ):
                        count += self.check_xmas(c, d)
        return count

    def check_x_mas(self, a_coord: Coord) -> bool:
        """Check to see for X-MAS."""
        diagonals: list[str] = [
            self.letter(a_coord + d)
            for d in (Coord(1, 1), Coord(1, -1), Coord(-1, -1), Coord(-1, 1))
        ]
        return sorted(diagonals) == ["M", "M", "S", "S"] and diagonals[0] != diagonals[2]

    def count_x_mas(self) -> int:
        """Count occurrences of X-MAS."""
        count = 0
        for x in range(self.width):
            for y in range(self.height):
                c = Coord(x, y)
                if self.letter(c) == "A":
                    count += self.check_x_mas(c)
        return count


if __name__ == "__main__":
    grid = LetterGrid(list(stdin.readlines()))
    print(grid.count_xmas())
    print(grid.count_x_mas())

File: rust/src/bin/aoc_2024_04.rs
// Advent of Code 2024 - Day 4

use grid::Grid;
use std::io;

fn read_puzzle() -> Grid<char> {
    let mut puzzle: Grid<char> = Grid::new(0, 0);
    for line in io::stdin().lines() {
        let line = line.unwrap();
        let row: Vec<char> = line.chars().collect();
        puzzle.push_row(row);
    }
    puzzle
}

fn apply_delta(
    (c_row, c_col): (usize, usize),
    (delta_row, delta_col): (i32, i32),
    (size_row, size_col): (usize, usize),
) -> Option<(usize, usize)> {
    let row: i32 = i32::try_from(c_row).unwrap() + delta_row;
    let col: i32 = i32::try_from(c_col).unwrap() + delta_col;
    let row_max: i32 = i32::try_from(size_row).unwrap();
    let col_max: i32 = i32::try_from(size_col).unwrap();
    if row >= 0 && row < row_max && col >= 0 && col < col_max {
        Some((usize::try_from(row).unwrap(), usize::try_from(col).unwrap()))
    } else {
        None
    }
}

fn check_xmas(puzzle: &Grid<char>, x: (usize, usize), delta: (i32, i32)) -> bool {
    let mut c = x;
    for letter in ['M', 'A', 'S'] {
        if let Some(c_delta) = apply_delta(c, delta, puzzle.size()) {
            c = c_delta;
            if puzzle.get(c_delta.0, c_delta.1) != Some(&letter) {
                return false;
            }
        } else {
            return false;
        }
    }
    true
}

const DIRECTIONS: [(i32, i32); 8] = [
    (0, 1),
    (1, 1),
    (1, 0),
    (1, -1),
    (0, -1),
    (-1, -1),
    (-1, 0),
    (-1, 1),
];

fn count_xmas(puzzle: &Grid<char>) -> i32 {
    let mut count: i32 = 0;
    for (x_coord, x_letter) in puzzle.indexed_iter() {
        if x_letter == &'X' {
            for d in DIRECTIONS {
                if check_xmas(puzzle, x_coord, d) {
                    count += 1;
                }
            }
        }
    }
    count
}

const DIAGONALS: [(i32, i32); 4] = [(1, 1), (1, -1), (-1, -1), (-1, 1)];

fn check_x_mas(puzzle: &Grid<char>, a: (usize, usize)) -> bool {
    let mut diagonals: Vec<char> = Vec::new();
    for d in DIAGONALS {
        if let Some((c_row, c_col)) = apply_delta(a, d, puzzle.size()) {
            if let Some(c_char) = puzzle.get(c_row, c_col) {
                diagonals.push(*c_char);
            }
        }
    }
    if diagonals.len() < 4 || diagonals[0] == diagonals[2] {
        return false;
    }
    diagonals.sort();
    diagonals == vec!['M', 'M', 'S', 'S']
}

fn count_x_mas(puzzle: &Grid<char>) -> i32 {
    let mut count: i32 = 0;
    for (a_coord, a_letter) in puzzle.indexed_iter() {
        if a_letter == &'A' && check_x_mas(puzzle, a_coord) {
            count += 1;
        }
    }
    count
}

fn main() {
    let puzzle = read_puzzle();
    println!("{}", count_xmas(&puzzle));
    println!("{}", count_x_mas(&puzzle));
}
